### PR TITLE
Add off-graph KV-cache cell reference: multi-sequence batching

### PR DESCRIPTION
**Pull request by kiymetakdemir**
**Summary**

 Adds CellReferenceCache to the eager KV-cache reference: many sequences over one pool of per-token cells. A cell holds one token's K/V plus that token's position and the set of sequences owning it, so a sequence needn't occupy a contiguous range and two sequences can share cells — a fork sets a second owner bit instead of copying K/V. The batch is flat on the token axis (B = 1).

  **Files**

  - extension/llm/cache/reference_cache.py — CellReferenceCache beside ContiguousReferenceCache, a separate class rather than a flag. Per-cell position and owner bitset; a token claims the lowest free cell, which keeps the read window tight. Visibility is a property of the cell, not the layout — query i attends cell j iff j is occupied, shares a sequence with i, and is no newer than i — so the mask is a broadcast compare over the cell metadata rather than a host loop, and the spec is always EXPLICIT. begin_step declares the per-token sequences and decides admission before the forward; since cells are interchangeable, a step that passes that check cannot then fail to allocate. The first layer of a step allocates and memoizes the cells and mask and the rest reuse them, so allocation happens exactly once and there is no commit; a spent declaration plus a served-layer set turn a skipped begin_step into an error rather than a silent reuse of the previous step's sequences. seq_cp/seq_rm/seq_keep are bitset operations and a cell frees only once its owner set empties. DYNAMIC sizing grows the pool by appending rows, so a cell keeps the index the plan holds. Multi-dimensional positions are rejected rather than guessed at.
  - extension/llm/cache/test_update_and_attend.py — ten cases.

  **Testing**

  Ten cases added to the existing unittest suite:

  - a single sequence matches the cacheless baseline, so the degenerate case is not special-cased
  - two sequences prefilled in one forward, each matching the baseline it would have produced alone
  - a batched decode step continuing both sequences
  - a fork sharing the trunk's cells without consuming any, with the branch attending the trunk's history
  - seq_rm freeing only the cells no other sequence still owns
  - admission failing in begin_step, before the forward
  - the step protocol: a forward that skips begin_step, a declared/actual token count mismatch, and a sequence verb not disguising
  a skipped declaration
  - growth keeping every cell's index and bytes, and STATIC agreeing with DYNAMIC
  
pytest extension/llm/cache/test_update_and_attend.py -q

### CODE REVIEW

**Comment by metascroy** [comment]
@claude review this code

File: extension/llm/cache/reference_cache.py [hunk 1]
```diff
@@ -20,14 +20,16 @@
 The cache places K/V and returns the history plus an ``AttendSpec`` (a mask *semantic*). The attend
 mechanism (``attend`` below) is applied by the op/backend from that spec.
 
-Scope for this initial slice: single sequence, contiguous placement, float KV.
+Two caches share the op: ``ContiguousReferenceCache`` (one sequence appended in
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py:23]: It can be a follow-up PR, but didn't we rename ContiguousReferenceCache as SequenceCache or something in C++ code?

File: extension/llm/cache/reference_cache.py [hunk 2]
```diff
@@ -171,6 +173,264 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+        self._declared = False  # a declaration is spent by the step it allocates
+        self._plan: Optional[_CellStepPlan] = None
+        self._served: Set[int] = set()
+
+    # -- runner face: admission, lifecycle, sequence verbs ------------------
+
+    def free_cells(self) -> int:
+        return self._pos.count(-1)
+
+    def can_extend(self, n: int = 1) -> bool:
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py:281]: What does can_extend mean in the context of multiple sequences?

Whether the total concatenated length is extendible?

Comment (kiymetakdemir) [reply extension/llm/cache/reference_cache.py:281]: Yes exactly, it checks if there are enough cells for this step's total tokens up to a capacity

File: extension/llm/cache/reference_cache.py [hunk 3]
```diff
@@ -171,6 +173,264 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+        bit = 1 << seq
+        for i in range(self._used_end):
+            if self._seq[i] & bit and self._in_range(self._pos[i], p0, p1):
+                self._seq[i] &= ~bit
+                if self._seq[i] == 0:
+                    self._pos[i] = -1
+        self._shrink()
+        self._invalidate_plan()
+
+    def seq_keep(self, keep: int) -> None:
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py]: Get rid of this for now. We can always extend API in future

File: extension/llm/cache/reference_cache.py [hunk 4]
```diff
@@ -171,6 +173,264 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+
+    cells: torch.Tensor  # [n_tok] long -- the cell each query token was given
+    mask: torch.Tensor  # [n_tok, read_len] bool -- true = attend
+
+
+@experimental(
+    "update_and_attend KV cache is experimental and may change without notice."
+)
+class CellReferenceCache:
+    """Per-cell KV history for several sequences sharing one pool.
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py:230]: The API here is leaving a lot of bookkeeping to the caller to flatten sequences.

Can we have a utility that flattens {seq_id: (token_id, pos_id)} to the parallel arrays seq_id, token_id, pos_id?

File: extension/llm/cache/reference_cache.py [hunk 5]
```diff
@@ -171,6 +173,264 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+        self._v[layer_id][:, :, cells, :] = v.to(self.config.dtype)
+        return (
+            self._k[layer_id][:, :, :read_len, :],
+            self._v[layer_id][:, :, :read_len, :],
+            AttendSpec(kind=MaskKind.EXPLICIT, mask=self._plan.mask),
+        )
+
+    # -- internals ----------------------------------------------------------
+
+    def _allocate(self, position: torch.Tensor, device: torch.device) -> _CellStepPlan:
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py]: Why is device here?

File: extension/llm/cache/reference_cache.py [hunk 6]
```diff
@@ -171,6 +173,264 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+                f"{position.shape[-1]}"
+            )
+        positions = position.reshape(-1).tolist()
+        if len(positions) != len(self._step_seq):
+            raise ValueError(
+                f"begin_step declared {len(self._step_seq)} tokens, "
+                f"the forward carries {len(positions)}"
+            )
+        cells = [
+            self._claim(pos, 1 << seq) for pos, seq in zip(positions, self._step_seq)
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py]: How are prefixes being shared here?

Isn't 1 << seq just having one owners.

If (pos, seq, tok) all match, are we identify shared cell? How do we do prefix sharing with batch prefill?

File: extension/llm/cache/reference_cache.py [hunk 7]
```diff
@@ -171,6 +173,264 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+            raise RuntimeError(
+                f"KV cache full: {len(seq_ids)} tokens need as many cells, "
+                f"{self.free_cells()} free"
+            )
+        self._step_seq = list(seq_ids)
+        self._declared = True
+        self._plan = None
+        self._served.clear()
+
+    def seq_cp(self, src: int, dst: int, p0: int = 0, p1: Optional[int] = None) -> None:
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py]: Are both p0/p1 definable if only prefix sharing is supported?

File: extension/llm/cache/reference_cache.py [hunk 8]
```diff
@@ -171,6 +173,264 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+
+    # -- runner face: admission, lifecycle, sequence verbs ------------------
+
+    def free_cells(self) -> int:
+        return self._pos.count(-1)
+
+    def can_extend(self, n: int = 1) -> bool:
+        return self.free_cells() >= n
+
+    def seq_len(self, seq: int) -> int:
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py]: There's inconsistent use of seq vs. seq_id in the code.

Let's consolidate on seq_id.

File: extension/llm/cache/reference_cache.py [hunk 9]
```diff
@@ -171,6 +173,264 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+
+    def seq_cp(self, src: int, dst: int, p0: int = 0, p1: Optional[int] = None) -> None:
+        """Give dst a claim on src's cells -- a fork that copies no K/V."""
+        src_bit, dst_bit = 1 << src, 1 << dst
+        for i in range(self._used_end):
+            if self._seq[i] & src_bit and self._in_range(self._pos[i], p0, p1):
+                self._seq[i] |= dst_bit
+        self._invalidate_plan()
+
+    def seq_rm(self, seq: int, p0: int = 0, p1: Optional[int] = None) -> None:
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py]: Document that None for p1 means until end?

File: extension/llm/cache/reference_cache.py [hunk 10]
```diff
@@ -171,6 +173,264 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+        """Drop seq's claim; a cell frees only once no sequence owns it."""
+        bit = 1 << seq
+        for i in range(self._used_end):
+            if self._seq[i] & bit and self._in_range(self._pos[i], p0, p1):
```
Comment (metascroy) [inline extension/llm/cache/reference_cache.py]: If we start with [0, 10] populated for seq0 and seq1, sharing cell_ids [0, 10], and we then do seq_rm on seq1 from [2, 5], what does that mean for what remains?

Is that a valid operation?

File: extension/llm/cache/test_update_and_attend.py [hunk 11]
```diff
@@ -258,6 +259,201 @@ def forward(self, q, k, v, position):
...
+            return self.model(x, pos, torch.arange(x.shape[1]))
+
+    def test_single_sequence_matches_baseline(self):
+        x = torch.randn(1, 5, self.hidden)
+        out = self._step(self._cache(), x, list(range(5)), [0] * 5)
+        torch.testing.assert_close(
+            out, self.model.reference_forward(x, torch.arange(5)), atol=1e-4, rtol=1e-4
+        )
+
+    def test_batched_sequences_match_separate_runs(self):
```
Comment (metascroy) [inline extension/llm/cache/test_update_and_attend.py:315]: I think these would be eaiser to read if we had the flatten utiltity mentioned above

**Comment by metascroy** [comment]
@claude review the code again, along with reviewer comments

File: extension/llm/cache/reference_cache.py (hunk 6 above)
Comment (kiymetakdemir) [reply extension/llm/cache/reference_cache.py]: Yes currently during write it is only one owner, and shared with seq_cp

File: extension/llm/cache/reference_cache.py (hunk 10 above)
Comment (kiymetakdemir) [reply extension/llm/cache/reference_cache.py]: That is valid, seq0 stays as it is, seq1 owns positions 0,1 and 5-10. One use case could be attention sink; keeping the first tokens and evicting the ones after sink.

File: extension/llm/cache/reference_cache.py (hunk 7 above)
Comment (kiymetakdemir) [reply extension/llm/cache/reference_cache.py]: I dropped p0, I guess it has no use case.

File: extension/llm/cache/reference_cache.py (hunk 8 above)
Comment (kiymetakdemir) [reply extension/llm/cache/reference_cache.py]: Done, also changed self._seq → self._owners

File: extension/llm/cache/reference_cache.py (hunk 5 above)
Comment (kiymetakdemir) [reply extension/llm/cache/reference_cache.py]: It was there to build the plan's tensors on the same device as the K/V, since both are used against the pools. Now dropped it, it takes the device from the pools themselves (self._k[0].device).

**Review by metascroy — APPROVED** [review]

File: extension/llm/cache/test_update_and_attend.py [hunk 12]
```diff
@@ -258,6 +261,266 @@ def forward(self, q, k, v, position):
...
+
+        tokens, positions, seq_ids, logits_indices = flatten_step(
+            {0: (a[:, :3], 0), 1: (b[:, :2], 0)}
+        )
+        cache.begin_step(seq_ids)
+        with REGISTRY.active(self.cache_key):
+            self.model(tokens, positions, logits_indices)
+
+        tokens, positions, seq_ids, logits_indices = flatten_step(
+            {0: (a[:, 3:], 3), 1: (b[:, 2:], 2)}
```
Comment (digantdesai) [inline extension/llm/cache/test_update_and_attend.py:355]: who will keep track of these positions?

Comment (kiymetakdemir) [reply extension/llm/cache/test_update_and_attend.py:355]: Right now the caller does. The cache could report it instead: next_pos(seq_id) = one past the highest position among the cells that sequence owns (0 if it owns none), so the call becomes flatten_step({0: (tok, cache.next_pos(0))}) and the caller stops tracking positions itself.

File: extension/llm/cache/reference_cache.py [hunk 13]
```diff
@@ -171,6 +173,323 @@ def _spec(q_len: int, total: int, device: torch.device) -> AttendSpec:
...
+        truncates it at position k.
+        """
+        self._check_seq_id(seq_id)
+        bit = 1 << seq_id
+        for i in range(self._used_end):
+            if self._owners[i] & bit and self._in_range(self._pos[i], p0, p1):
+                self._owners[i] &= ~bit
+                if self._owners[i] == 0:
+                    self._pos[i] = -1
+        self._shrink()
```
Comment (digantdesai) [inline extension/llm/cache/reference_cache.py:345]: I guess this will wind it back id self._owners == 0 (when the last seq_id is removed)? if yes, do we have a test? if no, can we add? For memory leaks.

File: extension/llm/cache/test_update_and_attend.py
Comment (digantdesai) [inline extension/llm/cache/test_update_and_attend.py:66]: can you remind me why position and logits_indices has to be in the forward sign?

Comment (kiymetakdemir) [reply extension/llm/cache/test_update_and_attend.py:66]: position is for RoPE and logits_indices for lm head

File: extension/llm/cache/reference_cache.py [hunk 14]
```diff
@@ -20,14 +20,16 @@
 The cache places K/V and returns the history plus an ``AttendSpec`` (a mask *semantic*). The attend
 mechanism (``attend`` below) is applied by the op/backend from that spec.
 
-Scope for this initial slice: single sequence, contiguous placement, float KV.
+Two caches share the op: ``ContiguousReferenceCache`` (one sequence appended in
+place) and ``CellReferenceCache`` (many sequences over a pool of per-token cells,
```
Comment (digantdesai) [inline extension/llm/cache/reference_cache.py:24]: did you consider cell size to be >1 token? This is in the context of the paged attn impls.

File: extension/llm/cache/reference_cache.py (hunk 13 above)
Comment (kiymetakdemir) [reply extension/llm/cache/reference_cache.py:345]: Yes _shrink walks used_end back past every trailing cell whose owners went empty. Now I added test_freeing_the_tail_shrinks_the_read_window
